pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -81,124 +90,259 @@ func (z *zedrouter) getNIBridgeConfig(
+			SharedLabels: port.SharedLabels,
+			IfName:       port.IfName,
 			IsMgmt:       port.IsMgmt,
 			MTU:          port.MTU,
-			DNSServers:   types.GetDNSServers(*z.deviceNetworkStatus, ifName),
-			NTPServers:   types.GetNTPServers(*z.deviceNetworkStatus, ifName),
-		},
+			DNSServers:   types.GetDNSServers(*z.deviceNetworkStatus, port.IfName),
+			NTPServers:   types.GetNTPServers(*z.deviceNetworkStatus, port.IfName),
+		})
 	}
+	return portConfigs
 }
 
-// Update NI status and set interface name of the selected uplink
-// referenced by a logical label.
-func (z *zedrouter) setSelectedUplink(uplinkLogicalLabel string,
-	status *types.NetworkInstanceStatus) error {
-	if status.PortLogicalLabel == "" {
-		// Air-gapped
-		status.SelectedUplinkLogicalLabel = ""
-		status.SelectedUplinkIntfName = ""
-		return nil
+// Update the selection of device ports matching the port label.
+func (z *zedrouter) updateNIPorts(status *types.NetworkInstanceStatus) (
+	changed bool, err error) {
+	var (
+		newPorts      []*types.NetworkPortStatus
+		newPortLabels []string
+		newNTPServers []net.IP
+	)
+	if status.NtpServer != nil {
+		// The NTP server explicitly configured for the NI.
+		newNTPServers = append(newNTPServers, status.NtpServer)
+	}
+	if status.PortLabel != "" {
+		newPorts = z.deviceNetworkStatus.LookupPortsByLabel(status.PortLabel)
+		for _, port := range newPorts {
+			newPortLabels = append(newPortLabels, port.Logicallabel)
+			if port.NtpServer != nil {
+				// The NTP server explicitly configured for the port.
+				newNTPServers = append(newNTPServers, port.NtpServer)
+			}
+			// NTP servers received via DHCP.
+			newNTPServers = append(newNTPServers, port.NtpServers...)
+		}
 	}
-	status.SelectedUplinkLogicalLabel = uplinkLogicalLabel
-	if uplinkLogicalLabel == "" {
-		status.SelectedUplinkIntfName = ""
-		// This is potentially a transient state, wait for DPC update
-		// and uplink probing eventually finding a suitable uplink port.
-		return fmt.Errorf("no selected uplink port")
-	}
-	port := z.deviceNetworkStatus.LookupPortByLogicallabel(uplinkLogicalLabel)
-	if port == nil {
-		err := fmt.Errorf("label of selected uplink (%s) does not match any port",
-			uplinkLogicalLabel)
-		// Wait for DPC update
-		return err
-	}
-	if port.InvalidConfig {
-		return fmt.Errorf("port %s has invalid config: %s", port.Logicallabel,
-			port.LastError)
-	}
-	ifName := port.IfName
-	status.SelectedUplinkIntfName = ifName
-	ifIndex, exists, _ := z.networkMonitor.GetInterfaceIndex(ifName)
-	if !exists {
-		// Wait for uplink interface to appear in the network stack.
-		return fmt.Errorf("missing uplink interface '%s'", ifName)
-	}
-	if status.IsUsingUplinkBridge() {
-		_, ifMAC, _ := z.networkMonitor.GetInterfaceAddrs(ifIndex)
-		status.BridgeMac = ifMAC
-	}
-	return nil
+	newNTPServers = generics.FilterDuplicatesFn(newNTPServers, netutils.EqualIPs)
+	changed = changed || !generics.EqualSets(status.Ports, newPortLabels)
+	status.Ports = newPortLabels
```

Comment:
Just a question, when the config for the NI only has one label, the "PortLabel", and here the status will have the array of Ports which is also labels, that means the NI status will have multiple labels?

Reply:
In the config, the port label can be logical, matching one port, or a shared label, potentially matching multiple ports.
Whereas in the status we report logical ports of attached ports (after the shared label is resolved).
If NI has only one port attached, this list will have only one entry with the logical label of the port.
I have renamed `newPortLabels` to `newPortLogicalLabels` to make it clear that we do not store shared labels in the `status.Ports` list.